### PR TITLE
Refactoring controllers

### PR DESCRIPTION
**Pull request by Gahmbit**

### CODE REVIEW

File: backend/controllers/Account.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
+const Snapshot = require("../models/Snapshot");
+// const express = require("express");
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]: remove

File: backend/controllers/Account.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+};
+
+const saveCurrentStats = (req, res) => {
+  getAccount(req.params.account).then((rsData) => {
+    if (!rsData) {
+      res.status(404).send("User not found, please try again");
+    } else {
+      canSnap(rsData)
+        .then((snappable) => {
+          // console.log(snappable);
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]: extra console.log

File: backend/controllers/Account.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  });
+};
+
+//HELPER FUNCTIONS
+async function getAccount(account) {
+  const userData = await fetch(
+    `https://apps.runescape.com/runemetrics/profile/profile?user=${account}&activities=20`
+  );
+  const userJSON = await userData.json();
+  // console.log(userJSON);
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]: extra console.log

File: backend/controllers/Account.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  if (userJSON.error) return null;
+  else return transformSnapshot(userJSON);
+}
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]:
```suggestion
  if (userJSON.error) {
       return null;
  }
  return transformSnapshot(userJSON);
}
```

File: backend/routes/Account.js [hunk 5]
```diff
@@ -1,122 +1,23 @@
...
-      // takeSnapshot(account);
-      console.log(
-        `GET request @ /accounts/${req.params.account}, from ${req.ip}`
-      );
-    })
-    .catch((err) => {
-      res.send(err);
-      console.log(err);
-    });
+  controller.getCurrentStats(req, res);
```
Comment (the-code-raccoon) [inline backend/routes/Account.js:10]: can probably just do this instead for all routes
```
router.get("/:account", controller.getCurrenStats);
```

File: backend/controllers/Account.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    timestamp: -1,
+  });
+  if (
+    (mySnapshots === null) |
+    (mySnapshots === undefined) |
+    (mySnapshots[0] === null)
+  ) {
+    return null;
+  }
+  console.log(mySnapshots);
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]: extra console.log

File: backend/controllers/Account.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    nameLower: account.nameLower,
+  }).sort({
+    timestamp: -1,
+  });
+  if (
+    (mySnapshots === null) |
+    (mySnapshots === undefined) |
+    (mySnapshots[0] === null)
+  ) {
+    return null;
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]: make it return `undefined`

File: backend/controllers/Account.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  }
+  console.log(mySnapshots);
+  return mySnapshots;
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]: if (
```suggestion
  if (mySnapshots?.length) {
    return mySnapshots;
  }
```

File: backend/controllers/Account.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    skills: skillsObj,
+  };
+  return snap;
+}
+
+async function canSnap(account) {
+  const latestSnap = await Snapshot.find({ nameLower: account.nameLower })
+    .sort({ timestamp: -1 })
+    .limit(1);
+  // await console.log(latestSnap);
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]: remove

File: backend/controllers/Account.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+  if (
+    latestSnap[0] === undefined ||
+    Date.now() - latestSnap[0].timestamp >= snapshotGap
+  ) {
+    // takeSnapshot(account);
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]: remove

File: backend/controllers/Account.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    return true;
+  } else return false;
+}
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]:
```suggestion
  return (
    !latestSnap[0] ||
    Date.now() - latestSnap[0].timestamp >= snapshotGap
  )
```

**Review by the-code-raccoon — COMMENTED** [review]
Put all the https error codes in a separate file for consts

File: backend/routes/Account.js (hunk 5 above)
Comment (Gahmbit) [reply backend/routes/Account.js:10]: Tried, does not work, Express requires it to be a callback

**Comment by Gahmbit** [comment]
- made all corrections
- created separate file for constant values (consts.js)

File: backend/controllers/Account.js [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    .catch((err) => {
+      console.error(err);
+      res.status(SERVER_ERROR).send("Database Error: Please Try Again!");
+    });
+};
+
+const getAllStats = (req, res) => {
+  getAccount(req.params.account)
+    .then((rsData) => {
+      if (!rsData || rsData.nameLower === null) {
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]:
```suggestion
      if (!rsData?.nameLower ) {
```

File: backend/controllers/Account.js [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+        .then((snappable) => {
+          if (snappable) {
+            res.status(CREATED);
+            takeSnapshot(rsData);
+            let timestamp = Date.now();
+            res.send(
+              `Saved ${rsData.name}'s data at ${new Date(
+                timestamp
+              ).toISOString()}`
+            );
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]:
```suggestion
            res.send(
              `Saved ${rsData.name}'s data at ${(new Date()).toISOString()}`
            );
```

File: backend/controllers/Account.js [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+//HELPER FUNCTIONS
+async function getAccount(account) {
+  const userData = await fetch(
+    `https://apps.runescape.com/runemetrics/profile/profile?user=${account}&activities=20`
+  );
+  const userJSON = await userData.json();
+  if (userJSON.error) {
+    return null;
+  }
+  return transformSnapshot(userJSON);
```
Comment (the-code-raccoon) [inline backend/controllers/Account.js]:
```suggestion
  if (!userJSON.error) {
    return transformSnapshot(userJSON);
  }
```

**Pull request merged**